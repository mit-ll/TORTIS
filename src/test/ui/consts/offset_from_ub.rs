// ignore-x86 FIXME: missing sysroot spans (#53081)

#![feature(const_raw_ptr_deref)]
#![feature(const_ptr_offset_from)]
#![feature(ptr_offset_from)]

#[repr(C)]
struct Struct {
    data: u8,
    field: u8,
}

pub const DIFFERENT_ALLOC: usize = {
    //~^ NOTE
    let uninit = std::mem::MaybeUninit::<Struct>::uninit();
    let base_ptr: *const Struct = &uninit as *const _ as *const Struct;
    let uninit2 = std::mem::MaybeUninit::<Struct>::uninit();
    let field_ptr: *const Struct = &uninit2 as *const _ as *const Struct;
    let offset = unsafe { field_ptr.offset_from(base_ptr) };
    offset as usize
};

pub const NOT_PTR: usize = {
    //~^ NOTE
    unsafe { (42 as *const u8).offset_from(&5u8) as usize }
};

pub const NOT_MULTIPLE_OF_SIZE: usize = {
    //~^ NOTE
    let data = [5u8, 6, 7];
    let base_ptr = data.as_ptr();
    let field_ptr = &data[1] as *const u8 as *const u16;
    let offset = unsafe { field_ptr.offset_from(base_ptr as *const u16) };
    offset as usize
};

fn main() {}
