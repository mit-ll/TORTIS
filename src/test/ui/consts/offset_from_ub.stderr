error: any use of this value will cause an error
  --> $SRC_DIR/libcore/ptr/mod.rs:LL:COL
   |
LL |           intrinsics::ptr_offset_from(self, origin)
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |           |
   |           ptr_offset_from cannot compute offset of pointers into different allocations.
   |           inside call to `std::ptr::<impl *const Struct>::offset_from` at $DIR/offset_from_ub.rs:19:27
   | 
  ::: $DIR/offset_from_ub.rs:13:1
   |
LL | / pub const DIFFERENT_ALLOC: usize = {
LL | |
LL | |     let uninit = std::mem::MaybeUninit::<Struct>::uninit();
LL | |     let base_ptr: *const Struct = &uninit as *const _ as *const Struct;
...  |
LL | |     offset as usize
LL | | };
   | |__-
   |
   = note: `#[deny(const_err)]` on by default

error: any use of this value will cause an error
  --> $SRC_DIR/libcore/ptr/mod.rs:LL:COL
   |
LL |           intrinsics::ptr_offset_from(self, origin)
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |           |
   |           a memory access tried to interpret some bytes as a pointer
   |           inside call to `std::ptr::<impl *const u8>::offset_from` at $DIR/offset_from_ub.rs:25:14
   | 
  ::: $DIR/offset_from_ub.rs:23:1
   |
LL | / pub const NOT_PTR: usize = {
LL | |
LL | |     unsafe { (42 as *const u8).offset_from(&5u8) as usize }
LL | | };
   | |__-

error: any use of this value will cause an error
  --> $SRC_DIR/libcore/ptr/mod.rs:LL:COL
   |
LL |           intrinsics::ptr_offset_from(self, origin)
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |           |
   |           exact_div: 1 cannot be divided by 2 without remainder
   |           inside call to `std::ptr::<impl *const u16>::offset_from` at $DIR/offset_from_ub.rs:33:27
   | 
  ::: $DIR/offset_from_ub.rs:28:1
   |
LL | / pub const NOT_MULTIPLE_OF_SIZE: usize = {
LL | |
LL | |     let data = [5u8, 6, 7];
LL | |     let base_ptr = data.as_ptr();
...  |
LL | |     offset as usize
LL | | };
   | |__-

error: aborting due to 3 previous errors

